pkg/shared/utils/containerrootfs/totempdirectory.go

```diff
@@ -0,0 +1,168 @@
+// Copyright © 2023 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package containerrootfs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+)
+
+type Rootfs interface {
+	Dir() string
+	Cleanup() error
+}
+
+func ToTempDirectory(ctx context.Context, src string) (Rootfs, error) {
+	cache, exists := GetCacheFromContext(ctx)
+	if exists && cache != nil {
+		return cache.ToTempDirectory(ctx, src)
+	}
+	return toTempDirectory(ctx, src)
+}
+
+type tempDirRootfs struct {
+	dir string
+}
+
+func (tdrf *tempDirRootfs) Dir() string {
+	return tdrf.dir
+}
+
+func (tdrf *tempDirRootfs) Cleanup() error {
+	if tdrf.dir == "" {
+		return nil
+	}
+
+	err := os.RemoveAll(tdrf.dir)
+	if err != nil {
+		return fmt.Errorf("unable to remove temp directory: %w", err)
+	}
+	return nil
+}
+
+func toTempDirectory(ctx context.Context, src string) (Rootfs, error) {
+	tdrf := &tempDirRootfs{}
+
+	tmpdir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return tdrf, fmt.Errorf("unable to create temp directory: %w", err)
+	}
+	tdrf.dir = tmpdir
+
+	err = ToDirectory(ctx, src, tmpdir)
+	if err != nil {
+		return tdrf, fmt.Errorf("unable to output squashed image to directory: %w", err)
+	}
```

Comment:
If we failed here, shouldn't we delete the tempdir?

Reply:
In the current implementation its expected that even if an error is returned defer rootfs.Cleanup() is expected to be called like:

```
rootfs, err := toTempDirectory(...)
defer rootfs.Cleanup()
if err != nil {
   ...
}
```

This does feel a bit weird though...

Reply:
since it's internal function I guess we can do whatever we feel like, but to me it feels like - if an `err` returned from a function, the function is responsible for cleaning and `roots` will be `nil`

Reply:
I found a pattern online for handling defer's within a function only if the function is not successful, for example:
```
func setupThing() (func() error, error) {
   successful := false
   
   file, err := os.OpenFile(...)
   if err != nil {
       return func() error {}, err
   }
   defer func() {
      if !successful {
          if err := file.Close(); err != nil {
              log.Errorf(err)
          }
      }
   }()
   
   file2, err := os.OpenFile(...)
   if err != nil {
        return func() error {}, err
   }
   defer func() {
      if !successful {
          if err := file2.Close(); err != nil {
              log.Errorf(err)
          }
      }
   }()
   
   successful = true
   return func() error {
          if err := file.Close(); err != nil {
              return err
          }
          if err := file2.Close(); err != nil {
              return err
          }
   }
}
```

If the function returns before reaching `successful = true` then the defer's in this function trigger and clean up the files and any error's are logged, otherwise its the callers responsibility to call the returned cleanup function.